### DISCUSSION TITLE
Change Content-Encoding for assets

### DISCUSSION THREAD

**Discussion by exreplay** [nuxt/nuxt#16457] [Questions]
## Question

Could there be something like a hook or a server middlewares which runs before the static one?

## Status

We are using the GCP to host a nuxt3 project. To be specific, we are using Google Cloud Run. We were thinking about using `brotli` for serving compressed `mjs` and `css` files.

The first step was installing the https://github.com/vbenjs/vite-plugin-compression plugin and add it to the vite plugins list.
```ts
import viteCompression from 'vite-plugin-compression';

export default defineNuxtConfig({
  vite: {
    plugins: [viteCompression({ algorithm: 'brotliCompress' })]
  }
});
```
That worked surprisingly well right out of the box.

Because we can't make changes at the network layer, i tried to change the `Content-Encoding` inside a `Server Middleware`. That is not possible, because the internal `static` middleware, which handles all the assets, ends the response chain. Therefore the `Server Middleware` never gets called.

I was playing around with the final `static.mjs` file which was generated by the build command and found a solution. I updated the `readAsset` and `serveStatic` function.

```js
async function readAsset (id, res) {
  let assetPath = resolve(mainDir, getAsset(id).path);

  if(assetPath.endsWith('.mjs') || assetPath.endsWith('.css')) {
    try {
      await promises.access(`${assetPath}.br`, constants.R_OK | constants.W_OK);
      assetPath = `${assetPath}.br`;
      res.setHeader("Content-Encoding", "br");
    } catch {}
  }

  return promises.readFile(assetPath);
}
```

```js
async function serveStatic(req, res) {
  // ...code before
  const contents = await readAsset(id, res);
  // ...code after
}
```

Basically what it does is to check if the requested asset is a `mjs` or a `css` file, checks if there is `brotli` version in the public folder and if it is, changes the `assetPath` and sets the correct header.

Currently i'm using `yarn patch` to change the `@nuxt/nitro` package to my needs which works perfectly fine.

## Additional information

This is my patch file,

```
diff --git a/dist/index.mjs b/dist/index.mjs
index 6bd0ceda62a6150696ea547ebbf1e21fab7b60a1..56727ff03a407297b12bcb8aa02fd4505054b6b6 100644
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -379,7 +379,7 @@ function staticAssets(context) {
   return virtual$1({
     "#static-assets": `export default ${JSON.stringify(assets, null, 2)};`,
     "#static": `
-import { promises } from 'fs'
+import { promises, constants } from 'fs'
 import { resolve } from 'pathe'
 import { dirname } from 'pathe'
 import { fileURLToPath } from 'url'
@@ -387,8 +387,18 @@ import assets from '#static-assets'
 
 const mainDir = dirname(fileURLToPath(globalThis.entryURL))
 
-export function readAsset (id) {
-  return promises.readFile(resolve(mainDir, getAsset(id).path))
+export async function readAsset (id, res) {
+  let assetPath = resolve(mainDir, getAsset(id).path);
+
+  if(assetPath.endsWith('.mjs') || assetPath.endsWith('.css')) {
+    try {
+      await promises.access(\`\${assetPath}.br\`, constants.R_OK | constants.W_OK);
+      assetPath = \`\${assetPath}.br\`;
+      res.setHeader("Content-Encoding", "br");
+    } catch {}
+  }
+
+  return promises.readFile(assetPath);
 }
 
 export function getAsset (id) {
diff --git a/dist/runtime/server/static.js b/dist/runtime/server/static.js
index 765c747912f1430bda49cd29f36950421bbebbb2..f5947c76b1b91fea2d9056a19ca7445943b01efa 100644
--- a/dist/runtime/server/static.js
+++ b/dist/runtime/server/static.js
@@ -81,6 +81,6 @@ async function serveStatic(req, res) {
     res.setHeader("Cache-Control", `max-age=${TWO_DAYS}, immutable`);
   }
 
-  const contents = await (0, _static.readAsset)(id);
+  const contents = await (0, _static.readAsset)(id, res);
   return res.end(contents);
 }
\ No newline at end of file
diff --git a/dist/runtime/server/static.mjs b/dist/runtime/server/static.mjs
index a5cd4b18722f1f7a5531b4cc6ad0946a7e035b64..073545b87ca14412740344c8ccc4a0792c78df51 100644
--- a/dist/runtime/server/static.mjs
+++ b/dist/runtime/server/static.mjs
@@ -53,6 +53,6 @@ export default async function serveStatic(req, res) {
   if (isBuildAsset) {
     res.setHeader("Cache-Control", `max-age=${TWO_DAYS}, immutable`);
   }
-  const contents = await readAsset(id);
+  const contents = await readAsset(id, res);
   return res.end(contents);
 }
```

which can be used by setting a `resolution` for the `@nuxt/nitro` package inside the `package.json`.

```json
{
  "resolutions": {
    "@nuxt/nitro": "patch:@nuxt/nitro-edge@npm:3.0.0-27436017.7b6252a#.yarn/patches/@nuxt-nitro-edge-npm-3.0.0-27436017.7b6252a-e9f1994cf9"
  }
}
```

This is what i found in the nuxt repo

### `readAsset` function
https://github.com/nuxt/framework/blob/15f84f5ba0c31bd909ad633174146c19573b2f56/packages/nitro/src/rollup/plugins/static.ts#L41

### `serveStatic` function
https://github.com/nuxt/framework/blob/15f84f5ba0c31bd909ad633174146c19573b2f56/packages/nitro/src/runtime/server/static.ts#L72

**Comment by qnp** [comment]
I think there could be two approaches:
1) Add a hook or a possibility to setup `readAsset` and `serveStatic` layers, as you suggest, and then write a community module
2) Add an option in Nuxt3 config for out-of-the-box gzip/brotli compression, without the need of a module which uses a new hook

I can make the dev, any guidance on this ?

**Reply by qnp** [reply]
For the record, we could do this in Nuxt 2 directly in nuxt.config.js:
```js
  render: {
    static: {
      setHeaders(res) {
        res.setHeader('X-Frame-Options', 'ALLOWALL');
        res.setHeader('Access-Control-Allow-Origin', '*');
        res.setHeader('Access-Control-Allow-Methods', 'GET');
        res.setHeader(
          'Access-Control-Allow-Headers',
          'Origin, X-Requested-With, Content-Type, Accept'
        );
      },
    },
  },
```

**Reply by exreplay** [reply]
Thanks for your response. I Also found that in nuxt2 there was a compression option https://nuxtjs.org/docs/configuration-glossary/configuration-render/#compressor. So something like this was possible:

```js
// nuxt.config.js
import shrinkRay from 'shrink-ray-current'

export default {
  render: {
    compressor: shrinkRay()
  }
```

So thinking about it, there are probably several options and flexibilities to solve that.

## Compressor option
Have something like a `compressor` option (nuxt2) which lets you completely replace the `static` middleware. This is the most flexible option and gives you complete control over how you want to server your assets but it's the hardest one to setup.

## `static.setHeaders` option
As you replied there is already a setHeaders option in nuxt2 which could also be used in nuxt3. This might be the easiest method to configure but there is no way to tell the middleware that it should pick up the compressed files from the public folder.

## Out of the box compression
Compression out of the box for gzip/brotli might be the most convenient. The problem is that if someone wants to use another compression algorithm there should be hooks, which allows that. Also the nuxt core team has to maintain the compression.

## Hook inside `readAsset`
Having a hook inside the `readAsset` function would give the oportunity to do something like i did with my patch. The hook could give you the `res` and the `assetPath` which then lets you decide how to set the headers or point to the right files by modifying the path.

I would go with the last option. This way it abstracts away what the `serveStatic` layer is already doing but gives you the full control over manipulating the response and asset path. Also this way you could easily setup a community module.

@qnp What do you think?

**Comment by qnp** [comment]
Clear recap. I completely agree in going with the last solution. It is way more flexible and easy to maintain.
Yet, I'm not certain about the "hook" terminology. The implementation you propose acts more like a *middleware* applied on the serveStatic or readAsset layer.

I bet a core team member could give some guidance on this.

**Reply by exreplay** [reply]
I agree with that the "hook" terminology is not right in that place. I also think that a middleware would also not be the right thing to do because those should be a layer on its own and end the response chain at any time. That means if you implement a middleware before, the `serveStatic` logic hasn't run and if you add a middleware afterwards the response already ended and the custom middleware never gets executed.

Maybe also a good idea would be to have a static config parameter in your `nuxt.config.ts` which lets you implement custom logic for the middleware and then gets served as a `virtual` like `#static`, `#paths` or `#static-assets`.

https://github.com/nuxt/framework/blob/15f84f5ba0c31bd909ad633174146c19573b2f56/packages/nitro/src/runtime/server/static.ts#L4

https://github.com/nuxt/framework/blob/15f84f5ba0c31bd909ad633174146c19573b2f56/packages/nitro/src/rollup/plugins/static.ts#L37

**Comment by qnp** [comment]
@danielroe I'm tagging you explicitly as this discussion has not received attention of a core member. I think this should be a great improvement for Nuxt3 (and the websites it powers) to leverage gzip/brotli compression.

**Reply by danielroe** [reply]
I think we should support this in nitropack 👍 Would very much love an issue as soon as we've fully migrated to https://github.com/unjs/nitropack and the repository is public.

**Reply by exreplay** [reply]
@danielroe Thanks for your answer. I will make an issue as soon as the nitropack repo is public.

**Reply by exreplay** [reply]
@danielroe I created an issue in the nitro package https://github.com/unjs/nitro/issues/69. If you need me to provide more context in the issue, let me know.

**Reply by danielroe** [reply]
Thank you!

**Comment by exreplay** [comment]
@qnp So i had some time hacking around and found another way which lets you extend the `readAsset` function. As i wrote in my answer above i tried to use a virtual.

I implemented it by passing a path to a mjs file which can export different config options for the static middleware to extend on.

### nuxt.config.ts
```ts
export default defineNuxtConfig({
  nitro: {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    static: './staticConfig.mjs'
  }
});
```

### staticConfig.mjs
```js
import { promises, constants } from 'fs';

export const maxAge = 60 * 60 * 24 * 365;

export async function hook(assetPath, res) {
  if (assetPath.endsWith('.mjs') || assetPath.endsWith('.css')) {
    try {
      await promises.access(`${assetPath}.br`, constants.R_OK | constants.W_OK);
      assetPath = `${assetPath}.br`;
      res.setHeader('Content-Encoding', 'br');
    } catch {}
  }

  return assetPath;
}
```

As you can see, it is now also possible to eg. export a `maxAge` const.

So as you can see below, i'm just reading the file which got passed to the nuxt config and outputting it in the `#static-config` virtual. Now it's possible to import it anywhere you need it.

### dist/index.mjs
```js
  return virtual$1({
    "#static-assets": `export default ${JSON.stringify(assets, null, 2)};`,
    "#static-config": readFileSync(resolve(context._nuxt.rootDir, context.static), 'utf8'),
    "#static": `
import { promises } from 'fs'
import { resolve } from 'pathe'
import { dirname } from 'pathe'
import { fileURLToPath } from 'url'
import assets from '#static-assets'
import { hook } from '#static-config'

const mainDir = dirname(fileURLToPath(globalThis.entryURL))

export async function readAsset (id, res) {
  let assetPath = resolve(mainDir, getAsset(id).path);

  if (hook) assetPath = await hook(assetPath, res);

  return promises.readFile(assetPath);
}

export function getAsset (id) {
  return assets[id]
}
`
  });
```

### dist/runtime/server/static.mjs
```js
// ...
import { maxAge } from '#static-config';
// ...
res.setHeader("Cache-Control", `max-age=${maxAge || TWO_DAYS}, immutable`);
// ...
```

This is my new patch file and the package.json:

### Patch
```
diff --git a/dist/index.mjs b/dist/index.mjs
index e2baf5fb165bba0236a657eaf533cabd4f58e8b2..d65cfdfa75f9b614a6f9e707ed98b012be0b5832 100644
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -386,17 +386,23 @@ function staticAssets(context) {
   }
   return virtual$1({
     "#static-assets": `export default ${JSON.stringify(assets, null, 2)};`,
+    "#static-config": readFileSync(resolve(context._nuxt.rootDir, context.static), 'utf8'),
     "#static": `
 import { promises } from 'fs'
 import { resolve } from 'pathe'
 import { dirname } from 'pathe'
 import { fileURLToPath } from 'url'
 import assets from '#static-assets'
+import { hook } from '#static-config'
 
 const mainDir = dirname(fileURLToPath(globalThis.entryURL))
 
-export function readAsset (id) {
-  return promises.readFile(resolve(mainDir, getAsset(id).path))
+export async function readAsset (id, res) {
+  let assetPath = resolve(mainDir, getAsset(id).path);
+
+  if (hook) assetPath = await hook(assetPath, res);
+
+  return promises.readFile(assetPath);
 }
 
 export function getAsset (id) {
diff --git a/dist/runtime/server/static.js b/dist/runtime/server/static.js
index e8bc80c22375cd089cadece5956ab97ec203c2dc..9b5206aa7902242f72618e8f4fb2ea94f8d30801 100644
--- a/dist/runtime/server/static.js
+++ b/dist/runtime/server/static.js
@@ -13,6 +13,8 @@ var _static = require("#static");
 
 var _paths = require("#paths");
 
+var _staticConfig = require("#static-config");
+
 const METHODS = ["HEAD", "GET"];
 const TWO_DAYS = 2 * 60 * 60 * 24;
 const STATIC_ASSETS_BASE = process.env.NUXT_STATIC_BASE + "/" + process.env.NUXT_STATIC_VERSION;
@@ -77,7 +79,7 @@ async function serveStatic(req, res) {
   }
 
   if (isBuildAsset) {
-    res.setHeader("Cache-Control", `max-age=${TWO_DAYS}, immutable`);
+    res.setHeader("Cache-Control", `max-age=${_staticConfig.maxAge || TWO_DAYS}, immutable`);
   }
 
-  const contents = await (0, _static.readAsset)(id);
+  const contents = await (0, _static.readAsset)(id, res);
diff --git a/dist/runtime/server/static.mjs b/dist/runtime/server/static.mjs
index 92c69cf95b4d63e18e9376865da6a59777b4b59d..599ba92fc0542c3f84f787387b4a01f714f25aab 100644
--- a/dist/runtime/server/static.mjs
+++ b/dist/runtime/server/static.mjs
@@ -2,6 +2,7 @@ import { createError } from "h3";
 import { withoutTrailingSlash, withLeadingSlash, parseURL } from "ufo";
 import { getAsset, readAsset } from "#static";
 import { buildAssetsDir } from "#paths";
+import { maxAge } from '#static-config';
 const METHODS = ["HEAD", "GET"];
 const TWO_DAYS = 2 * 60 * 60 * 24;
 const STATIC_ASSETS_BASE = process.env.NUXT_STATIC_BASE + "/" + process.env.NUXT_STATIC_VERSION;
@@ -51,7 +52,7 @@ export default async function serveStatic(req, res) {
     res.setHeader("Last-Modified", asset.mtime);
   }
   if (isBuildAsset) {
-    res.setHeader("Cache-Control", `max-age=${TWO_DAYS}, immutable`);
+    res.setHeader("Cache-Control", `max-age=${maxAge || TWO_DAYS}, immutable`);
   }
-  const contents = await readAsset(id);
+  const contents = await readAsset(id, res);
   return res.end(contents);

```

### package.json

```json
{
  "resolutions": {
    "@nuxt/nitro": "patch:@nuxt/nitro-edge@npm:3.0.0-27470397.9ebea90#.yarn/patches/@nuxt-nitro-edge-npm-3.0.0-27470397.9ebea90-c36da0f21b"
  }
}

```

**Reply by qnp** [reply]
Interesting experiment. I wonder if it will be the final solution if implemented in nitropack ? Let's discuss that when it's open.

**Comment by qnp** [comment]
https://github.com/unjs/nitro/pull/449
